text/0000-stabilize-arbitrary-self-types.md

```diff
@@ -103,6 +103,10 @@ The receiver chain is constructed as follows:
 
 Method resolution will be adjusted to take into account a receiver's receiver chain by taking looking through all impls for all types appearing in the receiver chain.
 
+> [@adetaylor] I think we need more detail here about how this interacts with `autoderef.rs` within the method resolution process. At every step of the (arbitrarily long) `Autoderef` chain, it feels like we'll now explore an (arbitrarily long) `Receiver` chain, which would result in an O(n^2) number of possible method candidates. That can't be OK.
```

Comment:
I agree that in practice this doesn't seem like it will be bad, but I have a strong suspicion we won't be allowed to submit something into rustc which has worst-case O(n^2) performance. I might be wrong.

> Are there types for which `Deref` and `Receiver` resolve to different types? 

I can't think of such a situation.

I agree that we don't like the sound of the blanket implementation, _but_ most of the concerns were about _later_ adding such a blanket implementation and the compatibility implications which might result. We could probably do so from the outset without concern.

I do quite like the idea of constraining things this way. Irrespective of whether it helps with the theoretical worst-case performance as above, it feels like we should consider this because it might result in less user surprise. User surprise is bad. Any cases where `Receiver` and `Deref` lead to different outcomes could be very confusing. Hmmm....